File: crates/heimq/tests/parity/raw_protocol.rs
use anyhow::{Context, Result};
use bytes::{Buf, BufMut, Bytes, BytesMut};
use heimq_protocol::messages::api_versions_request::ApiVersionsRequest;
use heimq_protocol::messages::api_versions_response::ApiVersionsResponse;
use heimq_protocol::messages::init_producer_id_request::InitProducerIdRequest;
use heimq_protocol::messages::init_producer_id_response::InitProducerIdResponse;
use heimq_protocol::messages::produce_request::{
    PartitionProduceData, ProduceRequest, TopicProduceData,
};
use heimq_protocol::messages::produce_response::ProduceResponse;
use heimq_protocol::messages::{ProducerId, TopicName};
use heimq_protocol::protocol::{Decodable, Encodable, StrBytes};
use heimq_protocol::records::{
    Compression, Record, RecordBatchEncoder, RecordEncodeOptions, TimestampType,
};
use std::io::{Read, Write};
use std::net::TcpStream;
use std::time::Duration;

const CLIENT_ID: &str = "heimq-parity-raw";
const SOCKET_TIMEOUT: Duration = Duration::from_secs(15);

pub struct RawKafkaClient {
    stream: TcpStream,
    correlation_id: i32,
}

impl RawKafkaClient {
    pub fn connect(bootstrap: &str) -> Result<Self> {
        let stream = TcpStream::connect(bootstrap)
            .with_context(|| format!("connect raw protocol client to {bootstrap}"))?;
        stream.set_read_timeout(Some(SOCKET_TIMEOUT))?;
        stream.set_write_timeout(Some(SOCKET_TIMEOUT))?;
        Ok(Self {
            stream,
            correlation_id: 1,
        })
    }

    pub fn api_versions(&mut self) -> Result<ApiVersionsResponse> {
        self.send(18, 0, &ApiVersionsRequest::default())
            .context("ApiVersions v0")
    }

    pub fn init_producer_id(&mut self) -> Result<(i64, i16)> {
        let mut request = InitProducerIdRequest::default();
        request.transactional_id = None;
        request.transaction_timeout_ms = 60_000;
        request.producer_id = ProducerId(-1);
        request.producer_epoch = -1;

        let response: InitProducerIdResponse =
            self.send(22, 0, &request).context("InitProducerId v0")?;
        if response.error_code != 0 {
            anyhow::bail!("InitProducerId returned error {}", response.error_code);
        }
        Ok((response.producer_id.0, response.producer_epoch))
    }

    pub fn produce_sequence(
        &mut self,
        topic: &str,
        producer_id: i64,
        producer_epoch: i16,
        base_sequence: i32,
    ) -> Result<i16> {
        let mut partition = PartitionProduceData::default();
        partition.index = 0;
        partition.records = Some(encode_idempotent_batch(
            producer_id,
            producer_epoch,
            base_sequence,
        )?);

        let mut topic_data = TopicProduceData::default();
        topic_data.name = TopicName(StrBytes::from_string(topic.to_string()));
        topic_data.partition_data = vec![partition];

        let mut request = ProduceRequest::default();
        request.transactional_id = None;
        request.acks = 1;
        request.timeout_ms = 15_000;
        request.topic_data = vec![topic_data];

        let response: ProduceResponse = self.send(0, 3, &request).context("Produce v3")?;
        let code = response
            .responses
            .first()
            .and_then(|topic| topic.partition_responses.first())
            .map(|partition| partition.error_code)
            .context("Produce response missing topic/partition response")?;
        Ok(code)
    }

    fn send<R, S>(&mut self, api_key: i16, api_version: i16, request: &R) -> Result<S>
    where
        R: Encodable,
        S: Decodable,
    {
        let correlation_id = self.correlation_id;
        self.correlation_id += 1;

        let payload = encode_request(api_key, api_version, correlation_id, request)?;
        self.stream.write_all(&payload)?;

        let mut len_buf = [0u8; 4];
        self.stream.read_exact(&mut len_buf)?;
        let len = i32::from_be_bytes(len_buf);
        if len < 4 {
            anyhow::bail!("invalid response frame length {len}");
        }
        let mut response_buf = vec![0u8; len as usize];
        self.stream.read_exact(&mut response_buf)?;

        let mut cursor = std::io::Cursor::new(response_buf);
        let response_correlation_id = cursor.get_i32();
        if response_correlation_id != correlation_id {
            anyhow::bail!(
                "correlation id mismatch: request {correlation_id}, response {response_correlation_id}"
            );
        }
        S::decode(&mut cursor, api_version).context("decode response")
    }
}

fn encode_request<R: Encodable>(
    api_key: i16,
    api_version: i16,
    correlation_id: i32,
    request: &R,
) -> Result<Vec<u8>> {
    let mut body = BytesMut::new();
    body.put_i16(api_key);
    body.put_i16(api_version);
    body.put_i32(correlation_id);
    body.put_i16(CLIENT_ID.len() as i16);
    body.put_slice(CLIENT_ID.as_bytes());
    request.encode(&mut body, api_version)?;

    let mut framed = BytesMut::new();
    framed.put_i32(body.len() as i32);
    framed.extend_from_slice(&body);
    Ok(framed.to_vec())
}

fn encode_idempotent_batch(
    producer_id: i64,
    producer_epoch: i16,
    base_sequence: i32,
) -> Result<Bytes> {
    let record = Record {
        transactional: false,
        control: false,
        partition_leader_epoch: 0,
        producer_id,
        producer_epoch,
        timestamp_type: TimestampType::Creation,
        timestamp: i64::from(base_sequence),
        sequence: base_sequence,
        offset: 0,
        key: Some(format!("key-{base_sequence}").into()),
        value: Some(format!("value-{base_sequence}").into()),
        headers: Default::default(),
    };

    let mut encoded = BytesMut::new();
    RecordBatchEncoder::encode(
        &mut encoded,
        &[record],
        &RecordEncodeOptions {
            version: 2,
            compression: Compression::None,
        },
    )?;
    Ok(encoded.freeze())
}

File: tests/conformance/python/conformance/raw_protocol.py
"""A minimal, hand-rolled Kafka protocol client.

Two workloads need to send frames no client library will send: a duplicate
idempotent sequence number, and one that skips ahead. That requires building
Produce requests by hand.

The Rust original built these frames with `heimq_protocol` -- the server's own
protocol library -- so a bug in that encoder would be invisible: both sides of
the comparison shared it. Encoding independently here is the point.

Implements only what the workloads use: ApiVersions v0, InitProducerId v0,
Produce v3, and a v2 RecordBatch.
"""

import socket
import struct
from typing import Any

from crc32c import crc32c

CLIENT_ID = b"heimq-conformance-raw"
SOCKET_TIMEOUT = 15.0


def _put_nullable_string(buf: bytearray, s: str | None) -> None:
    if s is None:
        buf += struct.pack(">h", -1)
    else:
        raw = s.encode()
        buf += struct.pack(">h", len(raw)) + raw


def _put_string(buf: bytearray, s: str) -> None:
    raw = s.encode()
    buf += struct.pack(">h", len(raw)) + raw


def _varint(value: int) -> bytes:
    """Zig-zag encoded signed varint, as used by the v2 record format."""
    value = (value << 1) ^ (value >> 63) if value < 0 else value << 1
    out = bytearray()
    while True:
        chunk = value & 0x7F
        value >>= 7
        if value:
            out.append(chunk | 0x80)
        else:
            out.append(chunk)
            return bytes(out)


class Reader:
    def __init__(self, buf: bytes):
        self.buf = buf
        self.pos = 0

    def _take(self, n: int) -> bytes:
        b = self.buf[self.pos : self.pos + n]
        if len(b) != n:
            raise EOFError("short read")
        self.pos += n
        return b

    def int16(self) -> int:
        return struct.unpack(">h", self._take(2))[0]

    def int32(self) -> int:
        return struct.unpack(">i", self._take(4))[0]

    def int64(self) -> int:
        return struct.unpack(">q", self._take(8))[0]

    def string(self) -> str | None:
        n = self.int16()
        if n < 0:
            return None
        return self._take(n).decode()


def encode_record_batch(
    producer_id: int, producer_epoch: int, base_sequence: int
) -> bytes:
    """One record, v2 batch, no compression -- mirrors encode_idempotent_batch."""
    key = f"key-{base_sequence}".encode()
    value = f"value-{base_sequence}".encode()
    timestamp = base_sequence

    record = bytearray()
    record += struct.pack(">b", 0)  # attributes
    record += _varint(0)  # timestamp delta
    record += _varint(0)  # offset delta
    record += _varint(len(key)) + key
    record += _varint(len(value)) + value
    record += _varint(0)  # header count
    record_with_len = _varint(len(record)) + bytes(record)

    # Everything the CRC covers: attributes through the record set.
    body = bytearray()
    body += struct.pack(">h", 0)  # attributes
    body += struct.pack(">i", 0)  # last offset delta
    body += struct.pack(">q", timestamp)  # base timestamp
    body += struct.pack(">q", timestamp)  # max timestamp
    body += struct.pack(">q", producer_id)
    body += struct.pack(">h", producer_epoch)
    body += struct.pack(">i", base_sequence)
    body += struct.pack(">i", 1)  # record count
    body += record_with_len

    crc = crc32c(bytes(body))

    after_batch_len = bytearray()
    after_batch_len += struct.pack(">i", 0)  # partition leader epoch
    after_batch_len += struct.pack(">b", 2)  # magic
    after_batch_len += struct.pack(">I", crc)
    after_batch_len += body

    batch = bytearray()
    batch += struct.pack(">q", 0)  # base offset
    batch += struct.pack(">i", len(after_batch_len))
    batch += after_batch_len
    return bytes(batch)


class RawKafkaClient:
    def __init__(self, bootstrap: str):
        host, port = bootstrap.rsplit(":", 1)
        self.sock = socket.create_connection((host, int(port)), timeout=SOCKET_TIMEOUT)
        self.sock.settimeout(SOCKET_TIMEOUT)
        self.correlation_id = 1

    def close(self) -> None:
        self.sock.close()

    def __enter__(self) -> "RawKafkaClient":
        return self

    def __exit__(self, *exc: Any) -> None:
        self.close()

    def _send(self, api_key: int, api_version: int, body: bytes) -> Reader:
        correlation_id = self.correlation_id
        self.correlation_id += 1

        header = bytearray()
        header += struct.pack(">hhi", api_key, api_version, correlation_id)
        header += struct.pack(">h", len(CLIENT_ID)) + CLIENT_ID

        frame = bytes(header) + body
        self.sock.sendall(struct.pack(">i", len(frame)) + frame)

        (length,) = struct.unpack(">i", self._read_exact(4))
        if length < 4:
            raise RuntimeError(f"invalid response frame length {length}")
        r = Reader(self._read_exact(length))
        got = r.int32()
        if got != correlation_id:
            raise RuntimeError(
                f"correlation id mismatch: request {correlation_id}, response {got}"
            )
        return r

    def _read_exact(self, n: int) -> bytes:
        chunks = []
        remaining = n
        while remaining:
            chunk = self.sock.recv(remaining)
            if not chunk:
                raise EOFError("connection closed")
            chunks.append(chunk)
            remaining -= len(chunk)
        return b"".join(chunks)

    def api_versions(self) -> int:
        r = self._send(18, 0, b"")
        return r.int16()  # error_code

    def init_producer_id(self) -> tuple[int, int]:
        body = bytearray()
        _put_nullable_string(body, None)  # transactional_id
        body += struct.pack(">i", 60_000)  # transaction_timeout_ms

        r = self._send(22, 0, bytes(body))
        r.int32()  # throttle_time_ms
        error_code = r.int16()
        if error_code != 0:
            raise RuntimeError(f"InitProducerId returned error {error_code}")
        producer_id = r.int64()
        producer_epoch = r.int16()
        return producer_id, producer_epoch

    def produce_sequence(
        self, topic: str, producer_id: int, producer_epoch: int, base_sequence: int
    ) -> int:
        records = encode_record_batch(producer_id, producer_epoch, base_sequence)

        body = bytearray()
        _put_nullable_string(body, None)  # transactional_id
        body += struct.pack(">h", 1)  # acks
        body += struct.pack(">i", 15_000)  # timeout_ms
        body += struct.pack(">i", 1)  # topic_data count
        _put_string(body, topic)
        body += struct.pack(">i", 1)  # partition_data count
        body += struct.pack(">i", 0)  # partition index
        body += struct.pack(">i", len(records)) + records

        r = self._send(0, 3, bytes(body))
        topic_count = r.int32()
        if topic_count < 1:
            raise RuntimeError("Produce response missing topic response")
        r.string()  # topic name
        partition_count = r.int32()
        if partition_count < 1:
            raise RuntimeError("Produce response missing partition response")
        r.int32()  # partition index
        return r.int16()  # error_code
